### PR TITLE
PADV-506: Fix/Improve CCX LTI configuration compatibility

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-506

## Description

This PR adds compatibility fix for CCXs and external reusable configurations to the LTI consumer XBlock. This also includes changes to previously added fixes for CCXs using the LTI consumer XBlock. This changes allow the LTI consumer XBlock to sync values from the main CCX LTI configuration model to all of it's children CCX LTI configurations and vice versa.

- [x] Remove previous CCX fixes.
- [x] Add LtiConfiguration sync_configurations method.
- [x] Add model_to_dict function to utils module.
- [x] Modify LtiConfiguration save method.
- [x] Include unit tests for all added and modified code.

## Testing

1. Install and configure xblock-lti-consumer
2. Create a course and add a LTI consumer XBlock.
3. Create a CCX of the previously created course. 
4. Launch the CCX course unit with the LTI consumer XBlock.
5. Go to http://localhost:18000/admin/lti_consumer/lticonfiguration/.
6. Both the main LTI configuration and it's children CCX configuration should have the same values.
7. Update any value on the main LTI configuration.
8. Go to the children CCX LTI configuration.
9. The values from the main LTI configuration should be reflected.
10. Delete both LTI configurations.
11. Launch the CCX course unit with the LTI consumer XBlock.
12. Launch the course unit with the main LTI consumer XBlock.
13. Go back to the LTI configuration admin and search for both newly created configurations.
14. Both LTI configurations values should be synced.

## Reviewers

- [ ] @squirrel18
- [ ] @alexjmpb

### CODE REVIEW

File: lti_consumer/models.py [hunk 1]
```diff
@@ -273,11 +260,37 @@ def clean(self):
...
-        # Only allow creation with a CCX location if CCX master configuration exists.
-        if not self.pk and is_ccx_location(self.location) and not self.ccx_master_configuration:
-            raise ValidationError(_('CCX master course LTI configuration is required to create.'))
+    def sync_configurations(self):
+        """Syncronize main/children configurations.
 
+        This method will synchronize the field values of main/children configurations.
+        On a configuration with a CCX location, it will copy the values from the main course configuration,
+        on a configuration with another type of location, it will try to query any children configuration and
+        update their fields using the current configuration values.
```
Comment (Squirrel18) [inline lti_consumer/models.py]: Please, change this doc-string according to our conversation. @kuipumu

File: lti_consumer/models.py [hunk 2]
```diff
@@ -273,11 +260,37 @@ def clean(self):
...
+                # Query child CCX configurations using location block ID and CCX namespace.
+                child_configs = LtiConfiguration.objects.filter(
+                    location__endswith=str(self.location).split('@')[-1],
+                ).filter(
+                    location__startswith=CCXBlockUsageLocator.CANONICAL_NAMESPACE,
+                ).exclude(id=self.pk)
+                # Copy fields to child CCX configurations.
+                child_configs.update(**model_to_dict(self, EXCLUDED_FIELDS))
+            except IndexError:
+                log.exception('Error querying child LTI configurations: location parse failed')
```
Comment (Squirrel18) [inline lti_consumer/models.py]: Please, be more explicit with this exception message. @kuipumu

File: lti_consumer/utils.py [hunk 3]
```diff
@@ -322,8 +322,29 @@ def check_token_claim(token, claim_key, expected_value=None, invalid_claim_error
...
+    if exclude is None:
+        exclude = []
+
+    try:
+        # Copy model object fields.
+        object_fields = copy.deepcopy(model_object.__dict__)
+
+        # Remove private fields.
+        for key in list(object_fields):
+            if key.startswith('_'):
```
Comment (Squirrel18) [inline lti_consumer/utils.py]:
```suggestion
            if key.startswith('_') or key in exluded_fields:
```

File: lti_consumer/utils.py [hunk 4]
```diff
@@ -322,8 +322,29 @@ def check_token_claim(token, claim_key, expected_value=None, invalid_claim_error
...
+def model_to_dict(model_object, exclude=None):
     """
-    Check if the block location is from a CCX.
+    Get dictionary from model object.
+
+    This function will create a copy of a model object in a dictionary,
+    with all private and excluded fields removed.
     """
-    return isinstance(location, CCXBlockUsageLocator)
+    if exclude is None:
```
Comment (Squirrel18) [inline lti_consumer/utils.py]:
```suggestion
    if not exclude:
```

**Review by Squirrel18 — CHANGES_REQUESTED** [review]

**Review by Squirrel18 — APPROVED** [review]

**Review by alexjmpb — APPROVED** [review]

**Pull request merged**